nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ftp/commands/FtpCommandSTOR.java

```diff
@@ -0,0 +1,233 @@
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.ProcessSessionFactory;
+import org.apache.nifi.processors.standard.ListenFTP;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.SocketException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class FtpCommandSTOR extends AbstractCommand {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FtpCommandSTOR.class);
+    private final AtomicReference<ProcessSessionFactory> sessionFactory;
+    private final CountDownLatch sessionFactorySetSignal;
+
+    public FtpCommandSTOR(AtomicReference<ProcessSessionFactory> sessionFactory, CountDownLatch sessionFactorySetSignal) {
+        this.sessionFactory = sessionFactory;
+        this.sessionFactorySetSignal = sessionFactorySetSignal;
+    }
+
+    /**
+     * Execute command.
+     */
+    public void execute(final FtpIoSession ftpSession, final FtpServerContext context, final FtpRequest request) {
+        try {
+            executeCommand(ftpSession, context, request);
+        } catch (FtpCommandException ftpCommandException) {
+            if (ftpCommandException.getSubId() == null) {
+                ftpSession.write(new DefaultFtpReply(ftpCommandException.getFtpReturnCode(), ftpCommandException.getBasicMessage()));
+            } else {
+                ftpSession.write(LocalizedDataTransferFtpReply.translate(ftpSession, request, context,
+                        ftpCommandException.getFtpReturnCode(),
+                        ftpCommandException.getSubId(),
+                        ftpCommandException.getBasicMessage(),
+                        ftpCommandException.getFtpFile()));
+            }
+        } finally {
+            ftpSession.resetState();
+            ftpSession.getDataConnection().closeDataConnection();
+        }
+    }
+
+    private void executeCommand(FtpIoSession ftpSession, FtpServerContext context, FtpRequest request)
+            throws FtpCommandException {
+
+        final String fileName = getArgument(request);
+
+        checkDataConnection(ftpSession);
+
+        final FtpFile ftpFile = getFtpFile(ftpSession, fileName);
+
+        checkWritePermission(ftpFile);
+
+        sendReturnCode150(ftpSession, context, request, ftpFile.getAbsolutePath());
+
+        final DataConnection dataConnection = openDataConnection(ftpSession, ftpFile);
+
+        transferData(dataConnection, ftpSession, context, request, ftpFile);
+    }
+
+    private String getArgument(final FtpRequest request) throws FtpCommandException {
+        final String argument = request.getArgument();
+        if (argument == null) {
+            throw new FtpCommandException(FtpReply.REPLY_501_SYNTAX_ERROR_IN_PARAMETERS_OR_ARGUMENTS, "STOR", null, null);
```

Comment:
This must be the subID because it is passed on to the ftpserver library which uses it as the key to look up the correct reply message which might be subject to translation.